C++ Interop: Basic C++ record (class/struct/union) import support

bricknerb opened the pull request:
Focus: Calling static C++ function defined in C++ classes.

Limitations:
* Ignores visibility (public / protected / private).
* No support for: dynamic classes, member methods, data members, declarations without definitions, importing inheritance.

Based on #5142.

C++ Interop Demo with a class:

```c++
// hello_world.h

namespace some_namespace {

class MyClass {
 public:
  static void hello_world();
};

}  // namespace some_namespace
```

```c++
// hello_world.cpp

#include "hello_world.h"

#include <cstdio>

namespace some_namespace {

void MyClass::hello_world() { printf("Hello World!\n"); }

}  // namespace some_namespace
```

```carbon
// main.carbon

library "Main";

import Cpp library "hello_world.h";

fn Run() -> i32 {
  Cpp.some_namespace.MyClass.hello_world();
  return 0;
}
```

```shell
$ clang -c hello_world.cpp
$ bazel-bin/toolchain/carbon compile main.carbon
$ bazel-bin/toolchain/carbon link hello_world.o main.o --output=demo
$ ./demo
Hello World!
```

Part of #5150.

jonmeow commented on toolchain/check/import_cpp.cpp:
```diff
@@ -70,6 +71,9 @@ static auto GenerateAst(Context& context, llvm::StringRef importing_file_path,
       clang::tooling::getClangStripDependencyFileAdjuster(),
       clang::tooling::FileContentMappings(), &diagnostics_consumer, fs);
+  // Remove link to the diagnostics consumer before its deletion.
+  ast->getDiagnostics().setClient(nullptr);
```
What here is intended to be part of other PRs, including #5149, and what's for this PR? I only see a single commit.

jonmeow replied:
Note, basing on other PRs is okay, but it needs to have each PR's contents as a separate commit so that only the specific commit relevant to this PR can be reviewed.

bricknerb replied:
I've rebased, created a commit for #5142 and forced pushed.
I hope that was your intention.

Let me know if there are easy ways to base a PR on other PRs that doesn't require artificially regenerating equivalent commits. Perhaps I should have rebased on the branch with the base commit?

jonmeow replied:
> I've rebased, created a commit for #5142 and forced pushed. I hope that was your intention.

Yes, I just needed commits I could review.

> Let me know if there are easy ways to base a PR on other PRs that doesn't require artificially regenerating equivalent commits. Perhaps I should have rebased on the branch with the base commit?

Yes, that's essentially what I do. Something like:

```
git switch -c branch1 trunk
git commit -m "First set of changes"
git switch -c branch2 branch1
git commit -m "Second set of changes"
```

bricknerb replied:
Makes sense.
I typically start with branch2 and only figure out I need branch1 while working in branch2 so this would probably not work well for my typical use case.

bricknerb committed:
When import C++ names, add the note on the ClangLookup() as well
Currently has no effect because we can't do lookup in classes, yet.

Part of #4666.

bricknerb committed:
C++ Interop: Basic C++ record (class/struct/union) import support
Focus: Calling static C++ function defined in C++ classes.

Limitations:
* Ignores visibility (public / protected / private).
* No support for: dynamic classes, member methods, data members, declarations without definitions, importing inheritance.

Based on #5142.

Part of #5150.

bricknerb committed:
Add function comments

bricknerb committed:
Add a TODO for a proper location

bricknerb committed:
Use trailing return type syntax

jonmeow commented on toolchain/check/import_cpp.cpp:
```diff
@@ -275,6 +274,133 @@ static auto ImportNamespaceDecl(Context& context,
+       .extern_library_id = SemIR::LibraryNameId::None,
+       .non_owning_decl_id = SemIR::InstId::None,
+       .first_owning_decl_id = class_decl_id},
+
```
There shouldn't be a blank line here.

```suggestion
```

bricknerb replied:
Done.

jonmeow commented on toolchain/check/import_cpp.cpp:
```diff
@@ -275,6 +274,133 @@ static auto ImportNamespaceDecl(Context& context,
+  ReplaceInstBeforeConstantUse(context, class_decl_id, class_decl);
+
+  // Build the `Self` type using the resulting type constant.
+  // TODO: Decide whether we actually need this.
```
`self_type_id` is used in a lot of different ways, e.g. you can see it being used as part of how access control is enforced. In other words, providing this means we can write code that handles both Carbon and C++ types with less special-casing. Maybe a good point of discussion if you think it shouldn't be provided for C++ types?

bricknerb replied:
SG, removed the TODO.

jonmeow commented on toolchain/check/import_cpp.cpp:
```diff
@@ -275,6 +274,133 @@ static auto ImportNamespaceDecl(Context& context,
+  name_scope.set_cpp_decl_context(clang_decl);
+
+  // Introduce `Self`.
+  // TODO: Decide whether we actually need this for C++ based classes.
```
See comments on your `self_type_id` TODO, I think similar applies here, although more for native Carbon code than for toolchain code.

bricknerb replied:
Done.

jonmeow commented on toolchain/check/import_cpp.cpp:
```diff
@@ -275,6 +274,133 @@ static auto ImportNamespaceDecl(Context& context,
+       .pattern_block_id = SemIR::InstBlockId::None,
+       .implicit_param_patterns_id = SemIR::InstBlockId::None,
+       .param_patterns_id = SemIR::InstBlockId::None,
+       // TODO: Consider supporting extern.
```
Can you elaborate what you mean by "supporting extern"? Do you mean using `extern` to force direct imports of C++ dependencies, or trying to fit C++'s `extern` into [Carbon's extern](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/declaring_entities.md#extern-and-extern-library)? I read this more as the latter, but aren't the keyword semantics actually pretty different, even though we've reused the name?

bricknerb replied:
My thinking was that we might want to have some C++ classes be configured as Carbon `extern` classes.
However, I see that we always set this to false in functions as well, so it makes sense to be consistent here.
Removed the TODO.

jonmeow commented on toolchain/check/import_cpp.cpp:
```diff
@@ -275,6 +274,133 @@ static auto ImportNamespaceDecl(Context& context,
+                                SemIR::NameId name_id,
+                                clang::CXXRecordDecl* clang_decl)
+    -> SemIR::InstId {
+  clang_decl = clang_decl->getDefinition();
```
Had you considered using a different variable here, to make it easier to see that `clang_decl` below isn't the same as the argument?

bricknerb replied:
Done.
FWIW, I wanted to make sure we don't accidentally reuse the declaration and that's why I overwrote it.

jonmeow commented on toolchain/check/import_cpp.cpp:
```diff
@@ -275,6 +274,133 @@ static auto ImportNamespaceDecl(Context& context,
+
+  auto [class_id, class_decl_id] =
+      BuildClassDefinition(context, parent_scope_id, name_id, clang_decl);
+  SetClassCompleteTypeWitnessId(context, class_id);
```
What made you choose to factor these separately?

bricknerb replied:
PTAL.

jonmeow commented on toolchain/check/import_cpp.cpp:
```diff
@@ -275,6 +274,133 @@ static auto ImportNamespaceDecl(Context& context,
+  return {class_id, class_decl_id};
+}
+
+// Sets the `complete_type_witness_id` field for the given class.
+static auto SetClassCompleteTypeWitnessId(Context& context,
+                                          SemIR::ClassId class_id) -> void {
+  // The class type is now fully defined. Set its object representation.
+  context.classes().Get(class_id).complete_type_witness_id =
+      AddInst<SemIR::CompleteTypeWitness>(
+          context,
+          // TODO: Consider having a proper location here.
+          Parse::NodeId::None,
+          {.type_id =
+               GetSingletonType(context, SemIR::WitnessType::SingletonInstId),
+           .object_repr_id =
+               GetStructType(context, SemIR::StructTypeFieldsId::Empty)});
+}
```
Should we be trying to share CheckCompleteClassType, e.g. by moving it from handle_class.cpp to a new class.h/cpp? That is to say, the object_repr will eventually need to be correct, and the vptr will need to be compatible too. Maybe the code should actually be shared?

bricknerb replied:
I've `CheckCompleteClassType()` with  `ComputeClassObjectRepresentation()` to share a bit more logic and moved all the implementation to `class.cpp`.
Also moved `TryGetAsClass()` there so it can be reused.
Added parameters to replace using the `Context` parsing state (stack fields). I hope this makes sense.

jonmeow commented on toolchain/check/import_cpp.cpp:
```diff
@@ -275,6 +274,133 @@ static auto ImportNamespaceDecl(Context& context,
+static auto ImportCXXRecordDecl(Context& context, SemIR::LocId loc_id,
+                                SemIR::NameScopeId parent_scope_id,
+                                SemIR::NameId name_id,
+                                clang::CXXRecordDecl* clang_decl)
```
Is it feasible to more frequently pass around `const&` instead of a mutable `*`?

bricknerb replied:
while `getDefinition()`, it returns a non const `CXXRecordDecl`, so it seems more accurate to pass a non const.
I think that in general, constness in Clang is broken so trying to use the const API in many cases would break.

jonmeow replied:
I understand Clang doesn't always get const-ness perfect, but my understanding is they're trying to improve. I think we should do our best to use const to reflect our intent, not what a Clang API happens to return.

You commented at https://github.com/carbon-language/carbon-lang/pull/5156#discussion_r2007976992 that you were concerned about accidentally overwriting this. Wouldn't using either `const&`, or at least `&`, address that risk?

I think we should probably use `const&` here if it reflects the intent and compiles, when Clang is using `const` well enough for our use-case.

bricknerb replied:
Given that `clang::SemaBase::LookupQualifiedName()` requires a non const `DeclContext *`, I think we need a non const pointer and not try to keep a const one and then find a way to convert it to a non const for calling `LookupQualifiedName()`.

jonmeow replied:
IIUC, that's for name lookup, not this specific code? We've discussed that the name lookup should be working based on clang::DeclID, not pointers. Would a TODO here to make this a const reference once the storage on `Class` is fixed make sense?

bricknerb replied:
Right, we use the decl as the cpp_decl_context that can be used later for lookup.
Added a TODO.

jonmeow commented on toolchain/check/import_cpp.cpp:
```diff
@@ -275,6 +274,133 @@ static auto ImportNamespaceDecl(Context& context,
+  auto class_decl = SemIR::ClassDecl{
+      .type_id = SemIR::TypeType::SingletonTypeId,
+      .class_id = SemIR::ClassId::None,
+      .decl_block_id = context.inst_blocks().AddDefaultValue()};
```
`AddDefaultValue` is really there to later reassign, for cases that can't determine the block up-front. Do you instead want `InstBlockId::Empty` here?

bricknerb replied:
Done.
Also removed using it in `BuildClassDefinition()`.

jonmeow commented on toolchain/check/import_cpp.cpp:
```diff
@@ -275,6 +274,133 @@ static auto ImportNamespaceDecl(Context& context,
+                                 SemIR::NameId name_id,
+                                 clang::CXXRecordDecl* clang_decl)
+    -> std::tuple<SemIR::ClassId, SemIR::InstId> {
+  auto [class_id, class_decl_id] =
+      BuildClassDecl(context, parent_scope_id, name_id);
+  auto& class_info = context.classes().Get(class_id);
+
+  // Track that this declaration is the definition.
+  CARBON_CHECK(!class_info.has_definition_started());
+  class_info.definition_id = class_decl_id;
+  class_info.scope_id = context.name_scopes().Add(
+      class_decl_id, SemIR::NameId::None, class_info.parent_scope_id);
+  class_info.body_block_id = context.inst_blocks().AddDefaultValue();
```
This looks very similar to code in handle_class.cpp. Had you considered sharing code?

bricknerb replied:
Done.

jonmeow commented:
Sorry for forgetting to publish. If someone else picks this up while I'm out, I'll in particular highlight that I think this PR should be sharing code more, where it's currently copying from other files. Thanks!

bricknerb committed:
Create a different variable for the record definition

bricknerb committed:
Remove blank line

bricknerb committed:
Remove TODO for reconsidering `self_type_id`

bricknerb committed:
Remove the TODO for reconsidering `Self`

bricknerb committed:
Don't set a default InstBlockId in class decl

bricknerb committed:
Extract common code for class definition between:
* `HandleParseNode(Context& context, Parse::ClassDefinitionStartId node_id)`
* `BuildClassDefinition(Context& context, SemIR::NameScopeId parent_scope_id, SemIR::NameId name_id, clang::CXXRecordDecl* clang_decl)`

bricknerb committed:
Avoid setting `body_block_id` to `AddDefaultValue()` as it seems to not have any impact on tests.

bricknerb committed:
Reuse the logic for setting `self_type_id`

bricknerb committed:
Remove `extern` support TODO

bricknerb committed:
* Create `ComputeClassObjectRepresentation()` that wraps `CheckCompleteClassType()` and move all its implementation to class.cpp. Use it in `handle_class.cpp` and `import_cpp.cpp`.
* Move `TryGetAsClass()` to class.h so it can be used in `class.cpp` and in `handle_class.cpp`.

jonmeow commented on toolchain/check/class.h:
```diff
@@ -0,0 +1,32 @@
+                          SemIR::InstId class_decl_id) -> SemIR::Class&;
+
+// Compute object representation for a fully defined class.
+auto ComputeClassObjectRepresentation(
```
This should say `ObjectRepr`; it's one of our idiomatic abbreviations: https://github.com/carbon-language/carbon-lang/blob/trunk/toolchain/docs/idioms.md#abbreviations-used-in-the-code-aka-carbon-abbreviation-decoder-ring

```suggestion
auto ComputeClassObjectRepr(
```

bricknerb replied:
Done.

jonmeow commented on toolchain/check/class.h:
```diff
@@ -0,0 +1,32 @@
+    Context& context, Parse::NodeId node_id, SemIR::ClassId class_id,
+    llvm::ArrayRef<SemIR::InstId> field_decls,
+    llvm::ArrayRef<SemIR::InstId> vtable_contents,
+    llvm::ArrayRef<SemIR::InstId> inst_block) -> void;
```
What is `inst_block` and where do the instructions come from? Can you please rename it to be more specific?

bricknerb replied:
You're right, it's the contents of an instruction block, not the block itself.
https://github.com/carbon-language/carbon-lang/blob/0631e181849d29055ec87e363123216d20bef463/toolchain/check/handle_class.cpp#L609

So `class_inst_block_contents`, or just `inst_block_contents` since "class" is pretty obvious?
WDYT?

jonmeow replied:
Is this the body of the definition? How about `body`?

bricknerb replied:
Done.

jonmeow commented on toolchain/check/class.cpp:
```diff
@@ -0,0 +1,269 @@
+      TryEvalInst(context, SemIR::InstId::None,
+                  SemIR::ClassType{.type_id = SemIR::TypeType::SingletonTypeId,
+                                   .class_id = class_id,
+                                   .specific_id = specific_id}));
```
I think this should be using `GetClassType`, see #5257 for similar fixes

bricknerb replied:
Done.

jonmeow commented on toolchain/check/class.h:
```diff
@@ -0,0 +1,32 @@
+// Builds the `Self` type using the resulting type constant.
+auto SetNewClassSelfTypeId(Context& context, SemIR::ClassId class_id) -> void;
+
+// Tracks that this declaration is the definition and introduce `Self`.
+auto TrackClassDefinition(Context& context, SemIR::ClassId class_id,
```
I'm not sure what "tracks" means. If I understand correctly, this is intended to handle some logic tied to starting the class definition, and "introduce `Self`" is referring to the `AddRequiredName` call. Maybe:
```suggestion
// Starts the class definition, adding `Self` to name lookup.
auto StartClassDefinition(Context& context, SemIR::ClassId class_id,
```

bricknerb replied:
Done.

jonmeow commented on toolchain/check/class.h:
```diff
@@ -0,0 +1,32 @@
+
+// Tracks that this declaration is the definition and introduce `Self`.
+auto TrackClassDefinition(Context& context, SemIR::ClassId class_id,
+                          SemIR::InstId class_decl_id) -> SemIR::Class&;
```
Can we be specific that this is supposed to be the definition?
```suggestion
                          SemIR::InstId definition_id) -> SemIR::Class&;
```

bricknerb replied:
Done.

jonmeow commented on toolchain/check/class.h:
```diff
@@ -0,0 +1,32 @@
+
+// Tracks that this declaration is the definition and introduce `Self`.
+auto TrackClassDefinition(Context& context, SemIR::ClassId class_id,
+                          SemIR::InstId class_decl_id) -> SemIR::Class&;
```
Rather than returning a reference to the class, could this take it as a parameter? We typically avoid returning references like this, outside of `Get` functions.

bricknerb replied:
Done.

jonmeow commented on toolchain/check/class.h:
```diff
@@ -0,0 +1,32 @@
+// Otherwise returns `nullptr`.
+auto TryGetAsClass(Context& context, SemIR::TypeId type_id) -> SemIR::Class*;
+
+// Builds the `Self` type using the resulting type constant.
+auto SetNewClassSelfTypeId(Context& context, SemIR::ClassId class_id) -> void;
```
I think this is actually just getting the `Self` type and assigning it, since it could be implemented using `GetClassType`. Maybe "Sets" is the right action here?

Also, what is "the resulting type constant" referring to? Maybe:

```suggestion
// Sets the `Self` type for the class.
auto SetClassSelfType(Context& context, SemIR::ClassId class_id) -> void;
```

bricknerb replied:
Done.

jonmeow commented on toolchain/check/class.h:
```diff
@@ -0,0 +1,32 @@
+auto TrackClassDefinition(Context& context, SemIR::ClassId class_id,
+                          SemIR::InstId class_decl_id) -> SemIR::Class&;
+
+// Compute object representation for a fully defined class.
```
```suggestion
// Computes the object representation for a fully defined class.
```

Style: "Function comments should be written with an implied subject of _This function_" https://google.github.io/styleguide/cppguide.html#Function_Comments

So, "This function computes the object ...", not "This function compute object ..."

bricknerb replied:
Done.

jonmeow commented on toolchain/check/class.cpp:
```diff
@@ -0,0 +1,269 @@
+    llvm::ArrayRef<SemIR::InstId> field_decls,
+    llvm::ArrayRef<SemIR::InstId> vtable_contents,
+    llvm::ArrayRef<SemIR::InstId> inst_block) -> void {
+  // The class type is now fully defined. Compute its object representation.
```
This comment may have made sense in the code you're moving, but now it seems confusing -- what ensured this code is fully defined, other than the function comment that said so?

I see it's remained in handle_class.cpp, which might be where this really belongs -- so removing this would clean up a confusing duplication.

```suggestion
```

bricknerb replied:
Done.

jonmeow commented on toolchain/check/import_cpp.cpp:
```diff
@@ -275,6 +275,106 @@ static auto ImportNamespaceDecl(Context& context,
+                                SemIR::NameId name_id,
+                                clang::CXXRecordDecl* clang_decl)
+    -> SemIR::InstId {
+  clang::CXXRecordDecl* clang_def = clang_decl->getDefinition();
```
```suggestion
  const clang::CXXRecordDecl* clang_def = clang_decl->getDefinition();
```

Similar to the above, maybe it'd be good to use `const`?

bricknerb replied:
Similarly, I don't think we can use a non const `DeclContext`.

jonmeow reviewed (commented):
Thanks for the refactoring!

bricknerb committed:
Rename `ComputeClassObjectRepresentation()` to `ComputeClassObjectRepr()`

bricknerb committed:
Rename `TrackClassDefinition()` to `StartClassDefinition()`

bricknerb committed:
Rename `class_decl_id` to `definition_id` in `StartClassDefinition()`

bricknerb committed:
Fix `ComputeClassObjectRepr()` comment

bricknerb committed:
Remove confusing comment from `ComputeClassObjectRepr()`

bricknerb committed:
Rename `SetNewClassSelfTypeId()` to `SetClassSelfType()` and improve its comment

bricknerb committed:
Change `StartClassDefinition()` to get `Class` by reference instead of getting `ClassId` and returning `Class` by reference

bricknerb committed:
Use `GetClassType()` in `SetClassSelfType()`

bricknerb committed:
Rename `inst_block` to `inst_block_contents` in `ComputeClassObjectRepr()`

jonmeow commented on toolchain/check/import_cpp.cpp:
```diff
@@ -275,6 +275,106 @@ static auto ImportNamespaceDecl(Context& context,
+                         // TODO: Set vtable.
+                         /*vtable_contents=*/{},
+                         // TODO: Set block.
+                         /*inst_block=*/{});
```
Note, don't forget to update cpp files to reflect renames in the h file.

bricknerb replied:
Thanks! Done.

bricknerb committed:
Added a TODO for changing `clang_decl` parameter in `ImportCXXRecordDecl()` to `const &`

bricknerb committed:
Rename `inst_block` parameter of `ComputeClassObjectRepr()` to `body` and update callers and callees

jonmeow commented:
Approving, not merging due to conflicts, but I expect those'll be straightforward to sort out.

bricknerb committed:
Merge branch 'main' into static2
